### PR TITLE
Hccl port conflict detection

### PR DESCRIPTION
**Pull request by samplise**
### What changes were proposed in this pull request?

DLRover will find available ports for all workers as HCCL ports.

### Why are the changes needed?

HCCL may use occupied ports and end with failure. 

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: dlrover/python/common/grpc.py [hunk 1]
```diff
@@ -104,6 +104,25 @@ def find_free_port_in_set(ports):
     raise RuntimeError(f"Fail to find a free port in {ports}")
 
 
+def find_free_port_for_hccl(start=60000, end=70000) -> int:
+    n_npu = 16
```
Comment (workingloong) [inline dlrover/python/common/grpc.py:108]: Replace the magic number with a constant value.

File: dlrover/python/elastic_agent/torch/training.py [hunk 2]
```diff
@@ -730,6 +734,37 @@ def stop_executor(self):
         """Shutdown the executor to save the checkpoint."""
         self._save_ckpt_executor.shutdown(wait=False)
 
+    def sync_training_ports(self):
```
Comment (workingloong) [inline dlrover/python/elastic_agent/torch/training.py:737]: It is better to implement a separate class for HPU processes.

Comment (samplise) [reply dlrover/python/elastic_agent/torch/training.py:737]: We need to release this function asap. We will refactor this in future pr.

File: dlrover/python/common/grpc.py [hunk 3]
```diff
@@ -104,6 +104,24 @@ def find_free_port_in_set(ports):
     raise RuntimeError(f"Fail to find a free port in {ports}")
 
 
+def find_free_port_for_hccl(start=60000, end=70000) -> int:
+    cur_start = start
+    logger.info(f"try to find available port for hccl from {start}")
```
Comment (BalaBalaYi) [inline dlrover/python/common/grpc.py:109]: ```logger.info(f"Try to find available port for hccl from {start} to {end}.")```
Notice: capital letter and full stop.

File: dlrover/python/common/constants.py [hunk 4]
```diff
@@ -301,3 +301,7 @@ class CheckpointConstant(object):
 class Accelerators(object):
     NVIDIA_GPU = "nvidia.com/gpu"
     ASCEND_NPU = "ascend-npu"
+
+
+class AscendConstants(object):
+    NPU_PER_NODE = 16
```
Comment (BalaBalaYi) [inline dlrover/python/common/constants.py:307]: Is this assumed maximum number of cards in a single machine?

File: dlrover/python/master/node/job_manager.py [hunk 5]
```diff
@@ -150,3 +150,8 @@ def handle_training_failure(
     def collect_node_heart_beat(self, node_type, node_id, timestamp):
         """Collect the heart beat message of nodes."""
         pass
+
+    @abstractclassmethod
+    def sync_node_training_port(self, node_id, port):
+        """Collect the heart beat message of nodes."""
```
Comment (BalaBalaYi) [inline dlrover/python/master/node/job_manager.py:156]: Annotations unchanged.

File: dlrover/python/master/node/dist_job_manager.py [hunk 6]
```diff
@@ -177,6 +179,7 @@ def __init__(
         )
         self._scaler: Scaler = job_scaler
         self._init_training_node_manager()
+        self._training_node_configure = TrainingNodeConfigure()
```
Comment (BalaBalaYi) [inline dlrover/python/master/node/dist_job_manager.py:182]: The filed 'self._training_node_configure' should be defined in their parent class 'JobManager' for all the sub class need this field.

File: dlrover/python/master/node/training_node.py [hunk 7]
```diff
@@ -359,3 +358,68 @@ def update_critical_node(self, critical_node_restarts):
...
+        self._recv_node_training_ports: Dict[int, int] = {}
+        self._node_training_port = 0
+        self._next_check_node_training_port = 0
+        self._n_node = 0
+
+    def set_node_num(self, num):
+        logger.info(f"set worker count: {num}")
+        self._n_node = num
+
+    def sync_node_training_port(self, node_id, port) -> SyncNodeTrainingPorts:
```
Comment (BalaBalaYi) [inline dlrover/python/master/node/training_node.py:381]: This method need more unit test to cover.

File: dlrover/python/master/node/training_node.py [hunk 8]
```diff
@@ -359,3 +358,68 @@ def update_critical_node(self, critical_node_restarts):
...
+            if port < self._next_check_node_training_port:
+                return SyncNodeTrainingPorts(
+                    training_port=0,
+                    next_check_port=self._next_check_node_training_port,
+                )
+            self._recv_node_training_ports[node_id] = port
+            if len(self._recv_node_training_ports) == self._n_node:
+                min_port = 0
+                max_port = 0
+                for _, recv_port in self._recv_node_training_ports.items():
```
Comment (BalaBalaYi) [inline dlrover/python/master/node/training_node.py:398]: Just sort the list and min = 1st max = last

File: dlrover/python/master/node/dist_job_manager.py (hunk 6 above)
Comment (samplise) [reply dlrover/python/master/node/dist_job_manager.py:182]: Training node configure is only valid in distributed mode. It is better to put it here.

**Review by BalaBalaYi — APPROVED** [review]
LGTM

File: dlrover/python/master/node/dist_job_manager.py (hunk 6 above)
Comment (BalaBalaYi) [reply dlrover/python/master/node/dist_job_manager.py:182]: Later a PR will fix these staff.

File: dlrover/python/master/node/job_manager.py (hunk 5 above)
Comment (BalaBalaYi) [reply dlrover/python/master/node/job_manager.py]: Will fix in next PR.

**Pull request merged**